### ISSUE TITLE
pipe to proc

### ISSUE THREAD

**Issue by aurium** [shelljs/shelljs#148]
Does shelljs have something to simulate "`command1 | command2`"?

A possible usage:
```javascript
echo(mylist.join('\n')).pipe('egrep "[0-9]"')
```

**Comment by olsonpm** [comment]
FWIW, I just cloned the repo with the same question in mind.  The following command
```
grep -ir "pipe" ./
```
the word pipe only showed up three times in src/test.js and once in the README.  My guess to this question then is "no"

**Comment by aurium** [comment]
Well, lets design it?

My `pipe` example method is not a good solution. This design asks the `pipe` parameter to be a shell command to be executes like `exec` does. How to use built-in `sed` and `grep`?

A solution may write pipe as a context object for other `shelljs` filter methods:
```javascript
ls('/tmp').pipe().exec('grep "[0-9]"')  // system grep
ls('/tmp').pipe().grep('[0-9]')         // built-in grep
```
May `pipe` be configured? If not `pipe` can be implemented as a getter and the `()` is useless.

### A simpler way:
All output by `shelljs` method must be a instance of `Piped` and that has `sed`, `grep`, `cat`, `to`, `toEnd`,  and `pipe` methods. This will result in a simpler jQuery like code:
```javascript
ls('/tmp').pipe('grep "[0-9]"')  // filter first output by any system command
ls('/tmp').grep('[0-9]')         // filter first output by built-in filter methods
ls('/tmp').grep('[0-9]').sed('-r', 's/\.[^.]+$//').cat() // list numbered files without extension
```

**Comment by bradparks** [comment]
@aurium - so from your experience, is there anyway to do something like this, as you were suggesting?
```
ls('/tmp').grep('[0-9]').sed('-r', 's/\.[^.]+$//').cat() // list numbered files without extension 
```

**Comment by aurium** [comment]
Yes @bradparks!
That is pretty easy... Please see this gist with an implementation example:
https://gist.github.com/aurium/c913c85c45725814bf17

**Comment by bradparks** [comment]
@aurium - sweet! thanks so much for taking the time to link to this... much appreciated!

**Comment by aurium** [comment]
:thumbsup:

**Comment by amacfie** [comment]
I'd say this is a pretty major missing feature

**Comment by ilaif** [comment]
:+1:

**Comment by nfischer** [comment]
Any work on a PR for this? I'm personally interested in this feature, and would be willing to help develop a PR

**Comment by chrishiestand** [comment]
@nfischer with [25 open PRs](https://github.com/shelljs/shelljs/pulls) how likely is a PR to be merged? Perhaps the project owners need help?

**Comment by nfischer** [comment]
@chrishiestand I agree it's a bit of an issue. It looks to me like the maintainer was in the process of setting this up to receive more help (making the shelljs organization), but it looks like he never finished. I think all we can do as a community is develop useful features in the mean time and try to resolve any issues we can.

Perhaps we could start a new thread discussing future maintenance of the project to assist @arturadib?

**Comment by nfischer** [comment]
I wanted to ping this thread to say that it looks like we might be able to start merging some PRs in the next couple weeks. So if someone wanted to take this feature on, I'd be very interested in reviewing it and possibly merging.

I can also help develop the PR, but probably won't be able to start work on it for a little while. @aurium are you still interested in working on this?

**Comment by ariporad** [comment]
Yeah, me and @nfischer can now merge PRs, so any new ones will be merged ASAP, and we'll try to to get through the backlog.

**Comment by nfischer** [comment]
Also, I'll be opening a PR soon that will allow for pipes! So stay tuned (I just have some bugs to fix first).

**Comment by nfischer** [comment]
So for my implementation, I opted to go with the syntax:

```Javascript
echo('foo'); // regular command
echo('foo').cat(); // echo foo | cat (which actually does the same thing)
echo('foo\nbar').grep('o').cat(); // echo -e 'foo\nbar' | grep 'o' | cat
```

The following commands can be on the **right** side of a pipe (almost anything can be on the left):

 - `cat`
 - `grep`
 - `sed`

Is this agreeable/intuitive to everyone? I still have some bugs to resolve, but after that I'll try to push the PR.

**Comment by amacfie** [comment]
Can `exec` be on the right too?

**Comment by aurium** [comment]
Nice @nfischer! _Can `exec` be on the right too?_ +1

**Comment by nfischer** [comment]
I will look into this. Thanks for mentioning

**Comment by myme** [comment]
`exec` would be really nice for portable ways of building e.g. uglified browserify bundles (if you want to opt out of `gulp`/`grunt`):

```
browserify entry.js | uglifyjs > bundle.js
```

It would be nice then to have access to the `.to` on the right-hand side as well. Either directly from `exec` or through `cat`:

```javascript
const bundler = exec('browserify entry.js').exec('uglifyjs');
bundler.to('bundle.js');
// or
bundler.cat().to('bundle.js');
```

In the latter case you probably want cat to be silent (or at least mute it through options), as you really don't want the `bundle.js` content dumped to stdout. I'm not really sure about the details though. It's a usecase that I would believe would be quite common and really nice to have.

**Comment by nfischer** [comment]
@myme thanks for providing a use case! I'll look into doing this.

This should be very easy to do for async `exec()`. I'll see if I can figure out a hack for synchronous `exec()` as well. If I can't get both, I'll do a PR for pipes for the utilities I've mentioned and ask for a contribution for `exec()` support.

**Comment by nfischer** [comment]
Marking this as blocked by #360, since I think it's cleanest to get that dealt with first

**Comment by nfischer** [comment]
I have an implementation of pipes out for review. If any of you all are interested in reviewing #370, please take a look. I even found a way to put `exec()` on the right-hand side of pipes, as requested.

Also, thanks to @aurium for the inspiration for how to design the pipes interface :+1:

**Comment by valessiobrito** [comment]
:clap: :clap: :clap: :clap:
